File: src/main/java/com/github/sblundy/elvish/psi/ElvishAssignmentBase.java
package com.github.sblundy.elvish.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class ElvishAssignmentBase extends ASTWrapperPsiElement implements ElvishAssignment, ElvishVariableScope {
    public ElvishAssignmentBase(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public @NotNull Collection<ElvishVariableDeclaration> findVariables(@NotNull String name, @NotNull List<String> ns) {
        List<ElvishVariableDeclaration> found = new ArrayList<>();
        for (ElvishVariableDeclaration declaration : getVariableList()) {
            if (declaration.nameMatches(name, ns)) {
                found.add(declaration);
            }
        }
        return found;
    }
}

File: src/main/kotlin/com/github/sblundy/elvish/psi/ElvishAssignmentBase.kt
package com.github.sblundy.elvish.psi

import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode

abstract class ElvishAssignmentBase(node: ASTNode) : ASTWrapperPsiElement(node), ElvishAssignment, ElvishVariableScope {
    override fun findVariables(ref: ReferenceWithNamespacePsiElement): Collection<ElvishVariableDeclaration> {
        return variableList.filter { declaration -> declaration.nameMatches(ref) }
    }

    override fun processVariables(processor: ElvishVariableScope.VariableProcessor) {
        variableList.any {
            processor.process(it)
        }
    }
}